File: frontend/src/types/flight.ts
export interface Flight {
  id: number;
  flight_number: string;
  airline_code: string;
  origin_airport: string;
  destination_airport: string;
  scheduled_departure: string;
  actual_departure: string;
  scheduled_arrival: string;
  actual_arrival: string;
  delay_minutes: number;
  delay_reason: string;
  distance: number;
  weather_condition: string;
  day_of_week: number;
  month: number;
  is_delayed: boolean;
}

export interface FlightStats {
  total_flights: number;
  delayed_count: number;
  delay_rate: number;
  avg_delay_minutes: number;
  ontime_rate: number;
}

export interface RouteDelay {
  route: string;
  origin: string;
  destination: string;
  total_flights: number;
  delayed_flights: number;
  delay_rate: number;
  avg_delay_minutes: number;
}

export interface WeatherDelay {
  weather: string;
  total: number;
  delayed: number;
  rate: number;
}

export interface TimeDelay {
  label: string;
  total: number;
  delayed: number;
  rate: number;
}

export interface PredictionRequest {
  airline: string;
  origin: string;
  destination: string;
  departure_datetime: string;
  weather_condition?: string;
}

export interface PredictionResult {
  delay_probability: number;
  estimated_delay_minutes: number;
  confidence: string;
  risk_band: string;
  severity_band: string;
  recommendations: string[];
  explanations: ExplanationItem[];
  error?: string;
  model_scores?: {
    xgboost: number;
    random_forest: number;
  };
}

export interface PredictionHistoryItem {
  id: number;
  airline: string;
  origin: string;
  destination: string;
  departure_datetime: string | null;
  delay_probability: number;
  estimated_delay_minutes: number;
  risk_band: string;
  severity_band: string;
  confidence: string;
  created_at: string | null;
}

export interface ApiResponse<T> {
  success: boolean;
  data: T;
  message: string;
  timestamp?: string;
}

export interface PaginatedData<T> {
  items: T[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export type PredictionHistoryResponse = PaginatedData<PredictionHistoryItem>;

export interface FlightListItem {
  id: number;
  flight_number: string;
  airline_code: string;
  origin: string;
  destination: string;
  scheduled_departure: string | null;
  weather: string | null;
  delay_minutes: number;
  is_delayed: boolean;
}

export interface FlightListResponse {
  flights: FlightListItem[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface ModelMetricSummary {
  accuracy?: number | null;
  precision?: number | null;
  recall?: number | null;
  f1_score?: number | null;
}

export interface FeatureImportanceItem {
  feature: string;
  importance: number;
}

export interface ExplanationItem {
  factor: string;
  impact: string;
  message: string;
}

export interface ModelInfo {
  status: string;
  xgboost?: ModelMetricSummary;
  random_forest?: ModelMetricSummary;
  ensemble?: ModelMetricSummary;
  feature_importance?: FeatureImportanceItem[];
  dataset_size?: number | null;
  train_size?: number | null;
  test_size?: number | null;
  trained_at?: string | null;
  accuracy?: number | null;
  precision?: number | null;
  recall?: number | null;
  f1_score?: number | null;
  features?: string[];
}

export interface UploadActionResult {
  filename?: string;
  rows_imported?: number;
  total_rows?: number;
  duration_seconds?: number;
  file_path?: string;
  type?: 'training';
  ensemble?: ModelMetricSummary;
}

File: backend/app/models/flight.py
"""
Flight data models - SQLAlchemy ORM + Pydantic schemas.
"""
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, ForeignKey
from sqlalchemy.orm import DeclarativeBase, relationship


# SQLAlchemy Base
class Base(DeclarativeBase):
    pass


# ── SQLAlchemy ORM Models ──────────────────────────────────────


class AirlineDB(Base):
    __tablename__ = "airlines"

    code = Column(String(5), primary_key=True)
    name = Column(String(100), nullable=False)

    flights = relationship("FlightDB", back_populates="airline")


class AirportDB(Base):
    __tablename__ = "airports"

    code = Column(String(5), primary_key=True)
    name = Column(String(100), nullable=False)
    city = Column(String(50))
    state = Column(String(50))
    latitude = Column(Float)
    longitude = Column(Float)


class FlightDB(Base):
    __tablename__ = "flights"

    id = Column(Integer, primary_key=True, autoincrement=True)
    flight_number = Column(String(10))
    airline_code = Column(String(5), ForeignKey("airlines.code"))
    origin_airport = Column(String(5), ForeignKey("airports.code"))
    destination_airport = Column(String(5), ForeignKey("airports.code"))
    scheduled_departure = Column(DateTime)
    actual_departure = Column(DateTime)
    scheduled_arrival = Column(DateTime)
    actual_arrival = Column(DateTime)
    delay_minutes = Column(Integer, default=0)
    delay_reason = Column(String(50))
    distance = Column(Float)
    weather_condition = Column(String(30))
    day_of_week = Column(Integer)
    month = Column(Integer)
    is_delayed = Column(Boolean, default=False)

    airline = relationship("AirlineDB", back_populates="flights")


class PredictionDB(Base):
    __tablename__ = "predictions"

    id = Column(Integer, primary_key=True, autoincrement=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    airline_code = Column(String(5))
    origin = Column(String(5))
    destination = Column(String(5))
    departure_datetime = Column(DateTime)
    delay_probability = Column(Float)
    estimated_delay_min = Column(Integer)
    confidence = Column(String(10))


# ── Pydantic Response Schemas ──────────────────────────────────


class FlightStatsResponse(BaseModel):
    total_flights: int
    delayed_count: int
    delay_rate: float
    avg_delay_minutes: float
    ontime_rate: float


class FlightListResponse(BaseModel):
    flights: List[dict]
    total: int
    page: int
    limit: int
    total_pages: int


class RouteDelayResponse(BaseModel):
    route: str
    origin: str
    destination: str
    total_flights: int
    delayed_flights: int
    delay_rate: float
    avg_delay_minutes: float
